### DISCUSSION TITLE
Compile C++ interface on multiple GPU cluster

### DISCUSSION THREAD

**Discussion by zezhong-zhang** [deepmodeling/deepmd-kit#2409] [Q&A] [answered]
The conda installed lammps could not use multiple GPUs on our small local cluster and large HPC properly, which I assume due to the conda uses its own MPI instead of the MPI from the HPC. Therefore, I am trying to compile all the C++ interface from the beginning on our local cluster, plug in the proper MPI during the lammps compilation. 

I followed the compile instruction here: https://docs.deepmodeling.com/projects/deepmd/en/latest/install/install-from-source.html

## system 
OS: Ubuntu 18.04.6 LTS x86_64 
CPU: Intel Xeon Silver 4114 (40) @ 2.201GHz 
GPU: NVIDIA TITAN V 

## Tensorflow C++ (finished correctly)
- version version 2.8
- Following https://docs.deepmodeling.com/projects/deepmd/en/latest/install/install-tf.2.8.html
- `mkdir -p tensorflow_root`, copied the libraries, export the path to the $tensorflow_root

## DeePMD-kit C++ (have error)
- version: current github master
- use `cmake -DTENSORFLOW_ROOT=$tensorflow_root -DCMAKE_INSTALL_PREFIX=$deepmd_root ..`
- The cmake installed is 3.10 (limited by the old ubuntu) instead of 3.16 suggested in the `CMakeLists.txt`, would this be an issue?

The error message

> -- Supported model version: 1.1 
> -- Found CUDA: /usr/local/cuda (found version "11.2") 
> -- Found CUDA in /usr/local/cuda, build nv GPU support
> -- Will not build AMD GPU support
> -- Enabled cpp interface build, looking for tensorflow_cc and tensorflow_framework
> CMake Error at cmake/Findtensorflow.cmake:217 (file):
>   file does not recognize sub-command GET_RUNTIME_DEPENDENCIES
> Call Stack (most recent call first):
>   CMakeLists.txt:134 (find_package)
> 
> 
> -- Found TensorFlow: /home/zezhong/data/tensorflow/tensorflow_root/include;/home/zezhong/data/tensorflow/tensorflow_root/include, /home/zezhong/data/tensorflow/tensorflow_root/lib/libtensorflow_cc.so, /home/zezhong/data/tensorflow/tensorflow_root/lib/libtensorflow_framework.so  in /home/zezhong/data/tensorflow/tensorflow_root/;/home/zezhong/data/tensorflow/tensorflow_root/;/usr/;/usr/local/ (found version "2.8.0")
> -- Automatically determined OP_CXX_ABI=1 
> -- Set GLIBCXX_USE_CXX_ABI=1
> CMake Error at cmake/Findtensorflow.cmake:415 (target_include_directories):
>   Cannot specify include directories for imported target
>   "TensorFlow::tensorflow_framework".
> Call Stack (most recent call first):
>   CMakeLists.txt:134 (find_package)
> 
> 
> CMake Error at cmake/Findtensorflow.cmake:417 (target_compile_definitions):
>   Cannot specify compile definitions for imported target
>   "TensorFlow::tensorflow_framework".
> Call Stack (most recent call first):
>   CMakeLists.txt:134 (find_package)
> 
> 
> CMake Error at cmake/Findtensorflow.cmake:441 (target_include_directories):
>   Cannot specify include directories for imported target
>   "TensorFlow::tensorflow_cc".
> Call Stack (most recent call first):
>   CMakeLists.txt:134 (find_package)
> 
> 
> CMake Error at cmake/Findtensorflow.cmake:443 (target_compile_definitions):
>   Cannot specify compile definitions for imported target
>   "TensorFlow::tensorflow_cc".
> Call Stack (most recent call first):
>   CMakeLists.txt:134 (find_package)
> 
> 
> -- Found OpenMP_C: -fopenmp  
> -- Found OpenMP_CXX: -fopenmp  
> -- Found OpenMP: TRUE   
> CMake Error at op/CMakeLists.txt:74 (target_precompile_headers):
>   Unknown CMake command "target_precompile_headers".
> 
> 
> -- Configuring incomplete, errors occurred!
> See also "/home/zezhong/data/deepmd-kit/source/build/CMakeFiles/CMakeOutput.log".
> See also "/home/zezhong/data/deepmd-kit/source/build/CMakeFiles/CMakeError.log".

## lammps 
Update: The DeePMD C++ is now compile, the error message above was due to the version of cmake.

I start compiling the lammps with CUDA-awareness, using the Nvidia HPC sdk 23.1, https://docs.nvidia.com/hpc-sdk/archive/23.1/
I followed the lammps build-in approach in https://docs.deepmodeling.com/projects/deepmd/en/master/install/install-lammps.html

The compilation was performed successfully until pppm_dplr.cpp, report couldn't find grid3d.h , with following error message:

> -DLAMMPS_VERSION_NUMBER=20230319 -I/home/zezhong/data/deepmd-kit/deepmd_root/include/ -DMPICH_SKIP_MPICXX -DOMPI_SKIP_MPICXX=1 -c ../pppm_dplr.cpp
> "../pppm_dplr.cpp", line 10: catastrophic error: cannot open source file "grid3d.h"
> #include "grid3d.h"
> ^
> 
> mpicxx -g -O3 -std=c++11 -DLAMMPS_GZIP -DLAMMPS_MEMALIGN=64 -std=c++14 -DHIGH_PREC -DLAMMPS_VERSION_NUMBER= -I/home/zezhong/data/deepmd-kit/deepmd_root/include/ -std=c++14 -DHIGH_PREC -DLAMMPS_VERSION_NUMBER=20230319 -I/home/zezhong/data/deepmd-kit/deepmd_root/include/ -DMPICH_SKIP_MPICXX -DOMPI_SKIP_MPICXX=1 -c ../fix_group.cpp
> 1 catastrophic error detected in the compilation of "../pppm_dplr.cpp".
> Compilation terminated.
> Makefile:114: recipe for target 'pppm_dplr.o' failed
> make[1]: *** [pppm_dplr.o] Error 2
> make[1]: *** Waiting for unfinished jobs....
> make[1]: Leaving directory '/media/hdd_3/zezhong/lammps-stable_23Jun2022_update3/src/Obj_mpi'
> Makefile:376: recipe for target 'mpi' failed
> make: *** [mpi] Error 2

**Comment by zezhong-zhang** [comment]
This is due to lammps version issue, solved by using the recent release. Also set `LAMMPS_VERSION_NUMBER` , see https://github.com/deepmodeling/deepmd-kit/issues/2275.

All the cpp files are compile but could not generate the `lmp_mpi`, any suggestions for why?
The error message: [error_mpi.log](https://github.com/deepmodeling/deepmd-kit/files/11045005/error_mpi.log)

**Reply by njzjz** [reply]
Did you run `make yes-user-deepmd`?

**Reply by njzjz** [reply]
Please share `Makefile.package` (under `src` directory) before and after executing `make yes-user-deepmd`.

**Reply by zezhong-zhang** [reply]
Yes I have run 
```
cp -r $deepmd_source_dir/source/build/USER-DEEPMD .
make yes-kspace
make yes-extra-fix
make yes-user-deepmd
make mpi -j8
```
The [Makefile.package](https://github.com/deepmodeling/deepmd-kit/files/11015506/Makefile.package.txt) after the `make yes-user-deepmd` is as attached, which has strangely no pkg in `PKG_INC`. I assume before the `make` is also empty.

**Reply by njzjz** [reply]
I can't reproduce this behavior, although I found another issue (#2414).

**Reply by zezhong-zhang** [reply]
I test to make `gpu` (without DeePMD) and `mpi` (with DeePMD) option from [recent lammps](https://github.com/lammps/lammps/releases/tag/stable_23Jun2022_update3 ), both now stop for undefined `ctz` and `ctzll`.

The compiler used is Nvidia HPC sdk 12.1 openmpi 4.0.5, with cuda and other libraries also in the HPC sdk, in an attempt to make the multi-gpu aware. 

The [Makefile.zip](https://github.com/deepmodeling/deepmd-kit/files/11031085/Makefile.zip) is as attached. 

Really appreciate your help!

### gpu or mpi
> make gpu -j 4 or make mpi -j 4
> Gathering installed package information (may take a little while)
> make[1]: Entering directory '/media/hdd_3/zezhong/lammps-stable_23Jun2022_update3/src'
> Gathering git version information
> make[1]: Leaving directory '/media/hdd_3/zezhong/lammps-stable_23Jun2022_update3/src'
> Compiling LAMMPS for machine gpu
> make[1]: Entering directory '/media/hdd_3/zezhong/lammps-stable_23Jun2022_update3/src/Obj_gpu'
> make[1]: Leaving directory '/media/hdd_3/zezhong/lammps-stable_23Jun2022_update3/src/Obj_gpu'
> make[1]: Entering directory '/media/hdd_3/zezhong/lammps-stable_23Jun2022_update3/src/Obj_gpu'
> mpicxx -g -O3 -std=c++11  -DLAMMPS_GZIP -std=c++14 -DLOW_PREC -DLAMMPS_VERSION_NUMBER=20220623 -I/home/zezhong/data/deepmd-kit/deepmd_root/include/   -std=c++14 -DLOW_PREC -DLAMMPS_VERSION_NUMBER=20220623 -I/home/zezhong/data/deepmd-kit/deepmd_root/include/   -std=c++14 -DLOW_PREC -DLAMMPS_VERSION_NUMBER=20220623 -I/home/zezhong/data/deepmd-kit/deepmd_root/include/   -DMPICH_SKIP_MPICXX -DOMPI_SKIP_MPICXX=1    -c ../fmtlib_format.cpp
> mpicxx -g -O3 -std=c++11  -DLAMMPS_GZIP -std=c++14 -DLOW_PREC -DLAMMPS_VERSION_NUMBER=20220623 -I/home/zezhong/data/deepmd-kit/deepmd_root/include/   -std=c++14 -DLOW_PREC -DLAMMPS_VERSION_NUMBER=20220623 -I/home/zezhong/data/deepmd-kit/deepmd_root/include/   -std=c++14 -DLOW_PREC -DLAMMPS_VERSION_NUMBER=20220623 -I/home/zezhong/data/deepmd-kit/deepmd_root/include/   -DMPICH_SKIP_MPICXX -DOMPI_SKIP_MPICXX=1    -c ../compute_reduce_region.cpp
> mpicxx -g -O3 -std=c++11  -DLAMMPS_GZIP -std=c++14 -DLOW_PREC -DLAMMPS_VERSION_NUMBER=20220623 -I/home/zezhong/data/deepmd-kit/deepmd_root/include/   -std=c++14 -DLOW_PREC -DLAMMPS_VERSION_NUMBER=20220623 -I/home/zezhong/data/deepmd-kit/deepmd_root/include/   -std=c++14 -DLOW_PREC -DLAMMPS_VERSION_NUMBER=20220623 -I/home/zezhong/data/deepmd-kit/deepmd_root/include/   -DMPICH_SKIP_MPICXX -DOMPI_SKIP_MPICXX=1    -c ../fix_nh.cpp
> "../fmt/format-inl.h", line 1952: error: identifier "ctz" is undefined
>     int t = ctz(n);
>             ^
> 
> "../fmt/format-inl.h", line 1980: error: identifier "ctzll" is undefined
>     int t = ctzll(n);
>             ^
> 
> 2 errors detected in the compilation of "../fmtlib_format.cpp".
> Makefile:114: recipe for target 'fmtlib_format.o' failed
> make[1]: *** [fmtlib_format.o] Error 2
> make[1]: *** Waiting for unfinished jobs....
> make[1]: Leaving directory '/media/hdd_3/zezhong/lammps-stable_23Jun2022_update3/src/Obj_gpu'
> Makefile:376: recipe for target 'gpu' failed
> make: *** [gpu] Error 2

**Reply by zezhong-zhang** [reply]
tested `make yes-user-deepmd` after `make yes-gpu` will actually empty the values (i.e. gpu) in `Makefile.package`
```
PKG_INC = -DLMP_GPU 
PKG_PATH = -L../../lib/gpu 
PKG_LIB = -lgpu 
```

**Reply by njzjz** [reply]
I fixed another issue in #2417.

To debug `USER-DEEPMD/Install.sh`, I suggest adding `set -vx` to the top of the file `USER-DEEPMD/Install.sh` and see what is printed.

**Reply by zezhong-zhang** [reply]
The log is quite long with a lot of repetitions, hence I only upload part of the log.. and I have to change `#$MY_DIR/.bin/$EXE "$@"` to `$MY_DIR/$EXE "$@"` in `/home/zezhong/data/hpc_sdk/Linux_x86_64/23.1/comm_libs/mpi/bin/env.sh` to start compile, otherwise it complains. The `make yes-user-deepmd` takes long time and fails eventually. 

[make_yes_user_deepmd.log](https://github.com/deepmodeling/deepmd-kit/files/11037930/make_yes_user_deepmd.log)

In comparison, if I don't use the openmpi, the  `make yes-user-deepmd`  finished quickly, the log is here 
[make_yes_user_deepmd_no_openmpi.log](https://github.com/deepmodeling/deepmd-kit/files/11038023/make_yes_user_deepmd_no_openmpi.log)
The Makefile.package indeed linked the deepmd library (though the layout is strange with large spacing)
[Makefile(no_openmpi).package.txt](https://github.com/deepmodeling/deepmd-kit/files/11038032/Makefile.no_openmpi.package.txt)

**Reply by njzjz** [reply]
I see. It looks like `source env.sh` is conflicted with OpenMPI `env.sh`.
The better way should be `source ./env.sh`.
cc @wanghan-iapcm

**Reply by zezhong-zhang** [reply]
- Indeed  manually `source ./env.sh` avoid the problem. But if I put it in the `Install.sh` it complains 
`Install.sh: 3: Install.sh: source: not found
`
- Also complains about the = operator. 

> if (test $1 = 1); then
> 
>         if (test -e ../Makefile.package); then
>                 sed -i -e "s|^PKG_INC =[ \t].*|& $NNP_INC|" ../Makefile.package
>                 sed -i -e "s|^PKG_PATH =[ \t].*|& $NNP_PATH|" ../Makefile.package
>                 Gsed -i -e "s|^PKG_LIB =[ \t].*|& $NNP_LIB|" ../Makefile.package
>         fi
> 
> elif (test $mode = 0); then
> 
>         if (test -e ../Makefile.package); then
>                 sed -i -e "s|$NNP_INC||g" ../Makefile.package
>                 sed -i -e "s|$NNP_PATH||g" ../Makefile.package
>                 sed -i -e "s|$NNP_LIB||g" ../Makefile.package
>         fi
> 
> fi
> + test = 1
> Install.sh: 41: test: =: unexpected operator
> + test = 0
> Install.sh: 49: test: =: unexpected operator

- ` sed -i -e "s|^PKG_INC =[ \t].*|& $NNP_INC|" ../Makefile.package` strangely put many space after `PKG_INC` which I don't see why. But it works fine otherwise. See  [Makefile.package.txt](https://github.com/deepmodeling/deepmd-kit/files/11044963/Makefile.package.txt)
- The compilation can proceed until `fmtlib_format.cpp` for undefined `ctz` and `ctzll`. See [make.log](https://github.com/deepmodeling/deepmd-kit/files/11044950/make.log)

I took OpenMPI as the MPICH is not yet part of the [MPI cuda aware list](https://developer.nvidia.com/blog/benchmarking-cuda-aware-mpi/), if successful, could this be an option for the conda recipe?

**Reply by njzjz** [reply]
You might have changed the shell from `bash` to `sh`. `=` and `source` and not supported in `sh`. The default shell is `bash` (see https://github.com/lammps/lammps/blob/fce1f8e0af22106abece97c8099815e97c8980c6/src/Makefile#LL3-L3C6) but I don't know if you changed it.

> The compilation can proceed until fmtlib_format.cpp for undefined ctz and ctzll. See [make.log](https://github.com/deepmodeling/deepmd-kit/files/11044950/make.log)

This should not be related to DeePMD-kit.

> I took OpenMPI as the MPICH is not yet part of the [MPI cuda aware list](https://developer.nvidia.com/blog/benchmarking-cuda-aware-mpi/),

To my knowledge, MVAPICH2 is ABI compatible with MPICH.

**Reply by zezhong-zhang** [reply]
> You might have changed the shell from bash to sh. = and source and not supported in sh. The default shell is bash (see https://github.com/lammps/lammps/blob/fce1f8e0af22106abece97c8099815e97c8980c6/src/Makefile#LL3-L3C6) but I don't know if you changed it.

I didn't change it in the file and the shell on the cluster is bash

> This should not be related to DeePMD-kit.

I put the `fmtlib_format.cpp` issue on https://github.com/lammps/lammps/issues/3700.

**Reply by zezhong-zhang** [reply]
I used the lammps-patch_8Feb2023 and put `source ./env.sh` in the `src/USER-DEEPMD/Install.sh`
```
cd lammps-patch_8Feb2023/src/
cp -r $deepmd_source_dir/source/build/USER-DEEPMD .
make yes-kspace
make yes-extra-fix
make yes-user-deepmd
```
The [Makefile.package](https://github.com/deepmodeling/deepmd-kit/files/11061507/Makefile.package.txt) now looks normal and the compilation finished correctly without error, see [make.log](https://github.com/deepmodeling/deepmd-kit/files/11061512/make.log). 

After adding `export OPAL_PREFIX=/home/zezhong/data/hpc_sdk/Linux_x86_64/23.1/comm_libs/mpi/` to the `.bashrc`,  the `lmp_mpi` generated do not complain but there is no output or screen print for some reason. Do you have any suggestions?

**Reply by zezhong-zhang** [reply]
I tried using cmake 
```
mkdir -p lammps-patch_8Feb2023/build/
cd lammps-patch_8Feb2023/build/
cmake -D PKG_PLUGIN=ON -D PKG_KSPACE=ON -D LAMMPS_INSTALL_RPATH=ON -D BUILD_SHARED_LIBS=yes -D CMAKE_INSTALL_PREFIX=${deepmd_root} -D CMAKE_INSTALL_LIBDIR=lib -D CMAKE_INSTALL_FULL_LIBDIR=${deepmd_root}/lib -D CMAKE_INSTALL_FULL_INCLUDEDIR=${deepmd_root}/include ../cmake
make
```

which generates the `lmp` successfully, see [cmake.log](https://github.com/deepmodeling/deepmd-kit/files/11063118/cmake.log).

But the issue is the same as the build-in mode, `./lmp -h` will freeze without any output or error. 
I tried `patchelf --set-rpath "${tensorflow_root}/lib" liblammps.so` which does not solve the problem.

**Reply by zezhong-zhang** [reply]
As mentioned in https://github.com/lammps/lammps/issues/3700, use `mpirun -n 1 ./lmp -h` can give the correct normal output.

however when I do `mpirun -n 1 lmp -in in.plugin.lammps `, it complains about undefined symbol. 

I took a look of https://github.com/conda-forge/lammps-feedstock/issues/123, build the lammps plug-in mode with 
`cmake -D PKG_PLUGIN=ON -D PKG_KSPACE=ON -D LAMMPS_INSTALL_RPATH=ON -D BUILD_SHARED_LIBS=yes -D CMAKE_INSTALL_PREFIX=${deepmd_root} -D CMAKE_INSTALL_LIBDIR=lib -D CMAKE_INSTALL_FULL_LIBDIR=${deepmd_root}/lib -D CMAKE_INSTALL_FULL_INCLUDEDIR=${deepmd_root}/include -D BUILD_LIB=ON -D BUILD_SHARED_LIBS=ON -D LAMMPS_INSTALL_RPATH=ON -D BUILD_MPI=ON -D PKG_MPIIO=ON -D LAMMPS_EXCEPTIONS=yes ../cmake`, the compilation finished smoothly, but running error report (undefined symbol) is the same...

```
> 
> --------------------------------------------------------------------------
> [[33869,1],0]: A high-performance Open MPI point-to-point messaging module
> was unable to find any relevant network interfaces:
> 
> Module: OpenFabrics (openib)
>   Host: ec-4029
> 
> Another transport will be used instead, although this may result in
> lower performance.
> 
> NOTE: You can disable this warning by setting the MCA parameter
> btl_base_warn_component_unused to 0.
> --------------------------------------------------------------------------
> LAMMPS (8 Feb 2023)
> OMP_NUM_THREADS environment is not set. Defaulting to 1 thread. (src/comm.cpp:98)
>   using 1 OpenMP thread(s) per MPI task
> Reading data file ...
>   triclinic box = (0 0 0) to (12.4447 12.4447 12.4447) with tilt (0 0 0)
>   1 by 1 by 1 MPI processor grid
>   reading atoms ...
>   192 atoms
>   read_data CPU = 0.002 seconds
> Plugin symbol lookup failure in file liblammps.so: 
> Open of file libdeepmd_lmp.so failed: /home/zezhong/data/deepmd-kit/deepmd_root/lib/libdeepmd_lmp.so: undefined symbol: _ZN3MPI8Datatype4FreeEv
> ERROR: Unrecognized pair style 'deepmd' (src/force.cpp:271)
> Last command: pair_style        deepmd frozen_model.pb
> --------------------------------------------------------------------------
> Primary job  terminated normally, but 1 process returned
> a non-zero exit code. Per user-direction, the job has been aborted.
> --------------------------------------------------------------------------
> --------------------------------------------------------------------------
> mpirun detected that one or more processes exited with non-zero status, thus causing
> the job to be terminated. The first process to do so was:
> 
>   Process name: [[33869,1],0]
>   Exit code:    1
> --------------------------------------------------------------------------
> 
```

**Reply by njzjz** [reply]
With `c++filt`, you can get the symbol name.
```sh
$ c++filt _ZN3MPI8Datatype4FreeEv
MPI::Datatype::Free()
```

I think you can check two things: for LAMMPS and the DeePMD-kit plugin, (1) whether the MPI is the same; (2) whether the LAMMPS code is the same.

**Reply by zezhong-zhang** [reply]
When I set `export LD_PRELOAD=/home/zezhong/data/hpc_sdk/Linux_x86_64/23.1/comm_libs/mpi/lib/libmpi_cxx.so`, the MPI error is gone. 

But now it cannot find `libdeepmd_op.so` in `${deepmd_root}/lib` even though it was explicitly added during the camke
```
cmake -D PKG_PLUGIN=ON -D PKG_KSPACE=ON -D LAMMPS_INSTALL_RPATH=ON -D BUILD_SHARED_LIBS=yes -D CMAKE_INSTALL_PREFIX=${deepmd_root} -D CMAKE_INSTALL_LIBDIR=lib -D CMAKE_INSTALL_FULL_LIBDIR=${deepmd_root}/lib -D CMAKE_INSTALL_FULL_INCLUDEDIR=${deepmd_root}/include -D BUILD_LIB=ON -D BUILD_MPI=ON -D PKG_MPIIO=ON -D LAMMPS_EXCEPTIONS=yes ../cmake
```
Adding the `${deepmd_root}/lib` to `LD_LIBRARY_PATH` does not work. It seems that somehow the lammps executable did not load the libraries for both the MPI and this error. 

Interestingly, the `libdeepmd_op.so` error was there in both the build-in mode and plugin mode, but the MPI error was only in the plug-in mode. 

```

LAMMPS (8 Feb 2023)
OMP_NUM_THREADS environment is not set. Defaulting to 1 thread. (src/comm.cpp:98)
  using 1 OpenMP thread(s) per MPI task
Reading data file ...
  triclinic box = (0 0 0) to (12.4447 12.4447 12.4447) with tilt (0 0 0)
  1 by 1 by 1 MPI processor grid
  reading atoms ...
  192 atoms
  read_data CPU = 0.002 seconds
Plugin symbol lookup failure in file liblammps.so: 
Loading plugin: deepmd pair style v2.2.1-dirty by Han Wang
Loading plugin: compute deeptensor/atom v2.2.1-dirty by Han Wang
Loading plugin: fix dplr v2.2.1-dirty by Han Wang
Loading plugin: kspace pppm/dplr v2.2.1-dirty by Han Wang
DeePMD-kit WARNING: Environmental variable TF_INTRA_OP_PARALLELISM_THREADS is not set. Tune TF_INTRA_OP_PARALLELISM_THREADS for the best performance. See https://deepmd.rtfd.io/parallelism/ for more information.
DeePMD-kit WARNING: Environmental variable TF_INTER_OP_PARALLELISM_THREADS is not set. Tune TF_INTER_OP_PARALLELISM_THREADS for the best performance. See https://deepmd.rtfd.io/parallelism/ for more information.
DeePMD-kit WARNING: Environmental variable OMP_NUM_THREADS is not set. Tune OMP_NUM_THREADS for the best performance. See https://deepmd.rtfd.io/parallelism/ for more information.
Summary of lammps deepmd module ...
  >>> Info of deepmd-kit:
  installed to:       /home/zezhong/data/deepmd-kit/deepmd_root
  source:             v2.2.1-dirty
  source branch:       master
  source commit:      3ac8c4c7
  source commit at:   2023-03-16 12:33:24 +0800
  surpport model ver.:1.1 
  build variant:      cuda
  build with tf inc:  /home/zezhong/data/tensorflow/tensorflow_root/include;/home/zezhong/data/tensorflow/tensorflow_root/include
  build with tf lib:  /home/zezhong/data/tensorflow/tensorflow_root/lib/libtensorflow_cc.so
  set tf intra_op_parallelism_threads: 0
  set tf inter_op_parallelism_threads: 0
  >>> Info of lammps module:
  use deepmd-kit at:  /home/zezhong/data/deepmd-kit/deepmd_root
  source:             v2.2.1-dirty
  source branch:      master
  source commit:      3ac8c4c7
  source commit at:   2023-03-16 12:33:24 +0800
  build float prec:   double
  build with tf inc:  /home/zezhong/data/tensorflow/tensorflow_root/include;/home/zezhong/data/tensorflow/tensorflow_root/include
  build with tf lib:  /home/zezhong/data/tensorflow/tensorflow_root/lib/libtensorflow_cc.so
DeePMD-kit WARNING: Environmental variable TF_INTRA_OP_PARALLELISM_THREADS is not set. Tune TF_INTRA_OP_PARALLELISM_THREADS for the best performance. See https://deepmd.rtfd.io/parallelism/ for more information.
DeePMD-kit WARNING: Environmental variable TF_INTER_OP_PARALLELISM_THREADS is not set. Tune TF_INTER_OP_PARALLELISM_THREADS for the best performance. See https://deepmd.rtfd.io/parallelism/ for more information.
DeePMD-kit WARNING: Environmental variable OMP_NUM_THREADS is not set. Tune OMP_NUM_THREADS for the best performance. See https://deepmd.rtfd.io/parallelism/ for more information.
ERROR: DeePMD-kit Error: libdeepmd_op.so is not found! You can add the library directory to LD_LIBRARY_PATH (/home/zezhong/data/deepmd-kit/source/lmp/pair_deepmd.cpp:933)
Last command: pair_style        deepmd frozen_model.pb

```

**Reply by njzjz** [reply]
You can set `export LD_DEBUG=libs` to debug libraries.

**Reply by zezhong-zhang** [reply]
Indeed, this is very helpful! From the libraries debug error as attached, it looks like the libraries can be found correctly but has error due to custom op similar to https://stackoverflow.com/questions/48189818/undefined-symbol-ztin10tensorflow8opkernele and https://github.com/tensorflow/tensorflow/issues/48866. I guess this is caused during the Tensorflow/DeePMD-kit compilation stage instead of lammps. I will try to rebuild Tensorflow/DeePMD-kit.

> 
> DeePMD-kit WARNING: Environmental variable OMP_NUM_THREADS is not set. Tune OMP_NUM_THREADS for the best performance. See https://deepmd.rtfd.io/parallelism/ for more information.
>      37915:     find library=libdeepmd_op.so [0]; searching
>      37915:      search path=/media/hdd_3/zezhong/hpc_sdk/Linux_x86_64/23.1/comm_libs/openmpi/openmpi-3.1.5/lib:/home/zezhong/data/deepmd-kit/deepmd_root/lib           (RUNPATH from file lmp)
>      37915:       trying file=/media/hdd_3/zezhong/hpc_sdk/Linux_x86_64/23.1/comm_libs/openmpi/openmpi-3.1.5/lib/libdeepmd_op.so
>      37915:       trying file=/home/zezhong/data/deepmd-kit/deepmd_root/lib/libdeepmd_op.so
>      37915:
>      37915:     find library=libdeepmd.so [0]; searching
>      37915:      search path=/media/hdd_3/zezhong/hpc_sdk/Linux_x86_64/23.1/comm_libs/openmpi/openmpi-3.1.5/lib:/home/zezhong/data/deepmd-kit/deepmd_root/lib           (RUNPATH from file lmp)
>      37915:       trying file=/media/hdd_3/zezhong/hpc_sdk/Linux_x86_64/23.1/comm_libs/openmpi/openmpi-3.1.5/lib/libdeepmd.so
>      37915:       trying file=/home/zezhong/data/deepmd-kit/deepmd_root/lib/libdeepmd.so
>      37915:
>      37915:     find library=libdeepmd_op_cuda.so [0]; searching
>      37915:      search path=/media/hdd_3/zezhong/hpc_sdk/Linux_x86_64/23.1/comm_libs/openmpi/openmpi-3.1.5/lib:/home/zezhong/data/deepmd-kit/deepmd_root/lib           (RUNPATH from file lmp)
>      37915:       trying file=/media/hdd_3/zezhong/hpc_sdk/Linux_x86_64/23.1/comm_libs/openmpi/openmpi-3.1.5/lib/libdeepmd_op_cuda.so
>      37915:       trying file=/home/zezhong/data/deepmd-kit/deepmd_root/lib/libdeepmd_op_cuda.so
>      37915:
>      37915:     /home/zezhong/data/deepmd-kit/deepmd_root/lib/libdeepmd_op.so: error: symbol lookup error: undefined symbol: _ZNK10tensorflow20OpKernelConstruction7HasAttrEN4absl12lts_2021110211string_viewE (fatal)
> ERROR: DeePMD-kit Error: libdeepmd_op.so is not found! You can add the library directory to LD_LIBRARY_PATH (/home/zezhong/data/deepmd-kit/source/lmp/pair_deepmd.cpp:933)
> Last command: pair_style        deepmd frozen_model.pb

**Reply by zezhong-zhang** [reply]
Okay, finally I compile the lammps with openmpi and the running of single gpu finished correctly. However, when I run on two gpus, it indeed calls two process on two individual card, but the calculation time is similar as on one gpu. When I checked the output [run_n2.log](https://github.com/deepmodeling/deepmd-kit/files/11153445/run_n2.log), others seems to be fine except 
```
[ec-4029:09460] 1 more process has sent help message help-mpi-btl-base.txt / btl:no-nics
[ec-4029:09460] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages
```
It seems like quite relevent to https://github.com/lammps/lammps/issues/3431 which also tested on openmpi. There is  https://github.com/lammps/lammps/pull/3432 but I am not sure how to apply to this case. 

I also tried to use mvmpich2 by compiling from the source, on single gpu is fine but on 2 gpus there has 

> Other MPI error, error stack: MPIR_Init_thread(493)....: MPID_Init(419)...........: channel initialization failed MPIDI_CH3_Init(688)......: MPIDI_CH3I_SMP_init(2221): MPIDI_CH3I_SMP_init(2216): process_vm_readv: Operation not permitted

**Reply by wanghan-iapcm** [reply, marked as answer]
You did not observe any acceleration by using 2 GPUs, because the system you tested (192 atoms) was too small to even use the computational power of 1 GPU. Please increase the size of your system to more than 1000 atoms (more than 500 atoms per GPU as rule of thumb) to check if you could achieve acceleration.

**Reply by zezhong-zhang** [reply]
Indeed! The system is too small to see the acceleration effect, I tried to run with 10k atoms and the parallelization is performed  properly, excited! The nvidia hpc sdk can do the compilation after setting things correctly. I think I can close the issue now, so much thanks to @njzjz @wanghan-iapcm!
